sync-core/src/main/java/com/cloudant/sync/query/IndexUpdater.java

```diff
@@ -0,0 +1,449 @@
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Future;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+class IndexUpdater {
+
+    private final SQLDatabase database;
+    private final Datastore datastore;
+
+    private final ExecutorService queue;
+
+    private static final Logger logger = Logger.getLogger(IndexUpdater.class.getName());
+
+    public IndexUpdater(SQLDatabase database, Datastore datastore, ExecutorService queue) {
+        this.datastore = datastore;
+        this.database = database;
+        this.queue = queue;
+    }
+
+    /**
+     *  Update all indexes in a set.
+     *
+     *  This indexes are assumed to already exist.
+     *
+     *  @param indexes Map of indexes and their definitions.
+     *  @param database The local database
+     *  @param datastore The local datastore
+     *  @param queue The executor service queue
+     *  @return index update success status (true/false)
+     */
+    public static boolean updateAllIndexes(Map<String, Object> indexes,
+                                           SQLDatabase database,
+                                           Datastore datastore,
+                                           ExecutorService queue) {
+        IndexUpdater updater = new IndexUpdater(database, datastore, queue);
+
+        return updater.updateAllIndexes(indexes);
+    }
+
+    /**
+     *  Update a single index.
+     *
+     *  This index is assumed to already exist.
+     *
+     *  @param indexName Name of index to update
+     *  @param fieldNames List of field names in the sort format
+     *  @param database The local database
+     *  @param datastore The local datastore
+     *  @param queue The executor service queue
+     *  @return index update success status (true/false)
+     */
+    public static boolean updateIndex(String indexName,
+                                      ArrayList<String> fieldNames,
+                                      SQLDatabase database,
+                                      Datastore datastore,
+                                      ExecutorService queue) {
+        IndexUpdater updater = new IndexUpdater(database, datastore, queue);
+
+        return updater.updateIndex(indexName, fieldNames);
+    }
+
+    @SuppressWarnings("unchecked")
+    private boolean updateAllIndexes(Map<String, Object> indexes) {
+        boolean success = true;
+
+        for (String indexName: indexes.keySet()) {
+            Map<String, Object> index = (Map<String, Object>) indexes.get(indexName);
+            ArrayList<String> fields = (ArrayList<String>) index.get("fields");
+            success = updateIndex(indexName, fields);
+            if (!success) {
+                break;
+            }
+        }
+
+        return success;
+    }
+
+    private boolean updateIndex(String indexName, ArrayList<String> fieldNames) {
+        boolean success;
+        Changes changes;
```

Comment:
Defining it in the do scope creates a compile error in the while check for (success && changes.size() > 0)